File: Medium/Longest Sub-Array with Sum K/longest-sub-array-with-sum-k.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
class Solution{
    public:
    int lenOfLongSubarr(int a[],  int n, int k) 
    { 
        // Complete the function
        unordered_map<long long, int> prevSum;
        long long sum=0;int maxLen=0;
        for(int i=0;i<n;i++){
            sum+=a[i];
            if(sum==k){
                maxLen=max(maxLen,i+1);
            }
            long long rem=sum-k;
            if(prevSum.find(rem)!=prevSum.end()){
                int len=i-prevSum[rem];
                maxLen=max(maxLen, len);
            }
            if(prevSum.find(sum)==prevSum.end()){
                prevSum[sum]=i;
            }
        }
        return maxLen;
    } 

};

//{ Driver Code Starts.

int main() {
	//code
	
	int t;cin>>t;
	while(t--)
	{
	    int n, k;
	    cin>> n >> k;
	    int a[n];
	    
	    for(int i=0;i<n;i++)
	        cin>>a[i];
	   Solution ob;
	   cout << ob.lenOfLongSubarr(a, n , k)<< endl;
	    
	}
	
	return 0;
}
// } Driver Code Ends

File: Difficulty: Medium/Longest Subarray with Sum K/longest-subarray-with-sum-k.java
// User function Template for Java

class Solution {
    public int longestSubarray(int[] arr, int k) {
        // code here
        Map<Integer,Integer> mp = new HashMap<>();
        int maxLen = 0;
        int sum = 0;
        
        for(int i=0;i<arr.length;i++){
            sum += arr[i];
            if(sum==k){
                maxLen = Math.max(maxLen,i+1);
            }
                int rem = sum - k;
                if(mp.containsKey(rem)==true){
                    int len = i - mp.get(rem);
                    maxLen=Math.max(maxLen,len);
                }
 
            if(mp.containsKey(sum)==false)mp.put(sum,i);
        }
        return maxLen;
    }
}


/*
    int lenOfLongSubarr(int arr[],  int n, int k) 
    { 
        // Complete the function
        int sum=0;
        int maxLen=0;
        unordered_map<int,int> mp;
        for(int i=0;i<n;i++){
            int len=0;
            sum+=arr[i];
            if(sum==k){
                len=i+1;
                maxLen=max(len,maxLen);
            }
            int remSum=sum-k;
            if(mp.find(remSum)!=mp.end()){
                len=i-mp[remSum];
                maxLen=max(maxLen,len);
            }
            if(mp.find(sum)==mp.end())mp[sum]=i;
        }
        return maxLen;
    } 
*/